### PR TITLE
build nightly casync build in jenkins

### PR DESCRIPTION
**Pull request by jnewb1**

### CODE REVIEW

File: selfdrive/updated/common.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import os
+import pathlib
+import subprocess
+
```
Comment (adeebshihadeh) [inline selfdrive/updated/common.py:4]: let's put all the new stuff in systemd/updated/ (no need to move the old one)

**Comment by jnewb1** [comment]
on master commits, in jenkins on device
  - creates a stripped version of OP
  - prebuilds
  - creates a casync release
  - uploads to bucket (need a token and azcopy on device?)

File: Jenkinsfile [hunk 2]
```diff
@@ -177,6 +177,15 @@ node {
 
     if (!env.BRANCH_NAME.matches(excludeRegex)) {
     parallel (
+      'release builds': {
+        if (env.BRANCH_NAME == 'master') {
+          // master channel
+          deviceStage("master channel", "tici-needs-can", [], [
```
Comment (adeebshihadeh) [inline Jenkinsfile:183]: let's start with nightly? don't want this to take unnecessary CI time yet

File: system/updated/common.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+def get_git_branch(path: str) -> str | None:
+  return run(["git", "rev-parse", "--abbrev-ref", "HEAD"], path).rstrip()
+
+
+def get_version(path: str) -> str:
+  with open(os.path.join(path, "common", "version.h")) as f:
+    return f.read().split('"')[1]
+
+
+def get_release_notes(path: str) -> str:
```
Comment (adeebshihadeh) [inline system/updated/common.py:21]: put in system/version.py?

File: system/updated/common.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  if env is None:
+    env = os.environ
+  return subprocess.check_output(cmd, cwd=cwd, stderr=subprocess.STDOUT, encoding='utf8', env=env)
+
+
+def get_git_branch(path: str) -> str | None:
+  return run(["git", "rev-parse", "--abbrev-ref", "HEAD"], path).rstrip()
+
+
+def get_version(path: str) -> str:
```
Comment (adeebshihadeh) [inline system/updated/common.py:16]: why is this copied from system/version.py?

File: system/updated/common.py [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+import subprocess
+
+
+def run(cmd: list[str], cwd: str = None, env = None) -> str:
+  if env is None:
+    env = os.environ
+  return subprocess.check_output(cmd, cwd=cwd, stderr=subprocess.STDOUT, encoding='utf8', env=env)
+
+
+def get_git_branch(path: str) -> str | None:
```
Comment (adeebshihadeh) [inline system/updated/common.py:12]: same here

File: release/create_prebuilt.sh [hunk 6]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/bash -e
+
+# run's on tici to create a prebuilt version of a release
```
Comment (adeebshihadeh) [inline release/create_prebuilt.sh:3]:
```suggestion
# runs on tici to create a prebuilt version of a release
```

File: release/create_prebuilt.sh [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+# Build
+export PYTHONPATH="$TARGET_DIR"
+scons -j$(nproc)
+
+# release panda fw
+CERT=/data/pandaextra/certs/release RELEASE=1 scons -j$(nproc) panda/
+
+# Cleanup
+find . -name '*.a' -delete
+find . -name '*.o' -delete
```
Comment (adeebshihadeh) [inline release/create_prebuilt.sh:26]: isn't this just a copy/pasted subset of another file?

**Review by adeebshihadeh — COMMENTED** [review]
The script names are not clear. for example, shouldn't "release/upload_casync_channel.sh" be "release/upload_casync_release.sh"?

File: release/create_prebuilt.sh (hunk 7 above)
Comment (jnewb1) [reply release/create_prebuilt.sh:23]: moved the common stuff into create_prebuilt and refactored build_release

File: system/version.py [hunk 8]
```diff
@@ -19,20 +19,19 @@ def cache(user_function: Callable[..., _RT], /) -> Callable[..., _RT]:
...
-def run_cmd_default(cmd: list[str], default: str = "") -> str:
+def run_cmd_default(cmd: list[str], default: str = "", cwd=None) -> str:
   try:
-    return run_cmd(cmd)
+    return run_cmd(cmd, cwd=cwd)
   except subprocess.CalledProcessError:
     return default
 
 
-@cache
```
Comment (jnewb1) [inline system/version.py:33]: need an uncached version of these for the updater. I also think caching these doesn't really gain anything and can have unforseen consequences

Comment (adeebshihadeh) [reply system/version.py:33]: it prevents touching the git repo too much... one of the reasons we're moving off of git

File: release/README.md [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+`channel` - a named version of openpilot (git branch, casync caidx) which receives updates<br>
+`prebuilt` - a channel which is already built for the comma 3/3x and contains only required files for running openpilot and identifying the channel<br>
+`release` - a `prebuilt` channel with `ALLOW_DEBUG=false` (`RELEASE=1` when building panda, ex: `nightly`, `release3`)<br>
+
+
+## openpilot channels
+
+| channel      | type        | description                                                       |
+| -----------  | ----------- | ----------                                                        |
+| release3     | `release`   | channel for end users                                             |
```
Comment (adeebshihadeh) [inline release/README.md:15]:
```suggestion
| release     | `release`   | stable release of openpilot                                             |
```

File: release/README.md [hunk 10]
```diff
@@ -0,0 +1,41 @@
+# openpilot releases
+
+
+## terms
+
+`channel` - a named version of openpilot (git branch, casync caidx) which receives updates<br>
```
Comment (adeebshihadeh) [inline release/README.md:6]: shouldn't need to use a `<br>` in a markdown file. just use bullets or something

File: release/upload_casync_channel.sh [hunk 11]
```diff
@@ -0,0 +1,9 @@
+#!/bin/bash
+
+CASYNC_DIR="${CASYNC_DIR:=/tmp/casync}"
+
+OPENPILOT_CHANNELS="https://commadist.blob.core.windows.net/openpilot-channels/"
```
Comment (adeebshihadeh) [inline release/upload_casync_channel.sh:5]: there's still too many things called "channels". conceptually, this is the "openpilot releases" bucket.

File: system/updated/casync/common.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  with open(path / ".caexclude", "w") as f:
+    # exclude everything except the paths already in the release
+    f.write("*\n")
+    f.write(".*\n")
+
+    for file in sorted(get_exclude_set(path)):
+      f.write(f"!{file}\n")
+
+
+def create_manifest(channel, version, release_notes):
```
Comment (adeebshihadeh) [inline system/updated/casync/common.py:41]: we already use manifest to refer to the file that describes the state of a version https://github.com/commaai/openpilot/blob/master/system/hardware/tici/agnos.json. this is more a version metadata info file

File: release/README.md [hunk 13]
```diff
@@ -0,0 +1,41 @@
...
+
+## openpilot channels
+
+| channel      | type        | description                                                       |
+| -----------  | ----------- | ----------                                                        |
+| release3     | `release`   | channel for end users                                             |
+| staging      | `release`   | staging channel for release3                                      |
+| nightly      | `release`   | generated nightly from last commit passing CI tests               |
+| master       | `prebuilt`  | current master commit with experimental features enabled          |
+| git branches | none        | installed manually, experimental features enabled, build required |
```
Comment (adeebshihadeh) [inline release/README.md:19]: git type?

File: release/README.md [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+# run on a tici, within the directory you want to create the channel from.
+# creates a prebuilt version of openpilot into BUILD_DIR and outputs the caidx
+# and other casync files into CASYNC_DIR for uploading to openpilot-channels.
+BUILD_DIR=/data/openpilot_build    \
+CASYNC_DIR=/data/casync            \
+OPENPILOT_CHANNEL=nightly          \
+release/create_casync_prebuilt.sh
+```
+
+`upload_casync_channel.sh` - helper for uploading a casync channel to `openpilot-channels`
```
Comment (adeebshihadeh) [inline release/README.md:36]: shouldn't this be upload_casync_release? "release" is like a git commit, and "channel" is analogous to a branch, so multiple channels can serve the same release

File: release/README.md [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+
+
+## release channel
+
+to create a release channel, set `RELEASE=1` environment variable when running build script
```
Comment (adeebshihadeh) [inline release/README.md:41]:
```suggestion
to create a release build, set `RELEASE=1` environment variable when running the build script
```

File: release/create_casync_channel.py [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+  parser.add_argument("channel", type=str, help="what channel this build is")
+  args = parser.parse_args()
+
+  target_dir = pathlib.Path(args.target_dir)
+  output_dir = pathlib.Path(args.output_dir)
+
+  create_manifest_file(target_dir, args.channel)
+  create_caexclude_file(target_dir)
+
+  digest, caidx = create_casync_channel(target_dir, output_dir, args.channel)
```
Comment (adeebshihadeh) [inline release/create_casync_channel.py:22]:
```suggestion
  digest, caidx = create_casync_release(target_dir, output_dir, args.channel)
```

**Review by adeebshihadeh — COMMENTED** [review]
I think there's still some terms usage to fix here.

Also maybe it makes sense to change "releases" to "builds"?

File: release/README.md [hunk 17]
```diff
@@ -0,0 +1,41 @@
+# openpilot releases
+
+
+## terms
+
+- `channel` - a named version of openpilot (git branch, casync caidx) which receives updates
+- `prebuilt` - a release which is already built for the comma 3/3x and contains only required files for running openpilot and identifying the release
+- `release` - a `prebuilt` release with `ALLOW_DEBUG=false` (`RELEASE=1` when building panda, ex: `nightly`, `release3`)
```
Comment (jnewb1) [inline release/README.md:8]:
```suggestion
- `debug` - a `prebuilt` release with `ALLOW_DEBUG=true` (`RELEASE=0` when building panda, ex: `nightly`, `release3`) which allows for testing experimental features
```

[the remark this answers is not in the document]
Comment (jnewb1) [reply release/README.md:8]:
```suggestion
- `debug` - a `prebuilt` release with `ALLOW_DEBUG=true` (`RELEASE=0` when building panda, ex: `nightly`, `release3`) which allows for testing experimental features
```

**Pull request merged**